In [ ]:
#Importando pandas
import pandas as pd

#Bibliotecas sklearn
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.linear_model import LogisticRegression

#Importando numpy
import numpy as np

# ***Exercício 1***

In [3]:
#Criando Dataframe
df = pd.read_csv('/content/sample_data/energydata_complete.csv', sep=',')
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

## Utlizando Regressão Linear

In [18]:
#Regressao linear
lr = LinearRegression()

#Preparando Dados
x = df.drop(['date', 'lights', 'Appliances'], axis=1)
y = df['Appliances']

#Dividir Dados
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

#Normalizar os Dados
scaler = MinMaxScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

#Treinar o modelo
lr.fit(x_train, y_train)

#Avaliar o modelo
y_pred = lr.predict(x_test)
r2 = r2_score(y_test, y_pred)

#R², RMSE e MAE
print('R²:', r2)
print('RMSE:', np.sqrt(mean_squared_error(y_test, y_pred)))
print('MAE:', mean_absolute_error(y_test, y_pred))

R²: 0.14890246319303535
RMSE: 93.64046093998029
MAE: 53.642977655849585


## Utlizando Árvore de Regressão

In [21]:
#Árvore de regressão
rf = RandomForestRegressor()

#Preparando Dados
x = df.drop(['date', 'lights', 'Appliances'], axis=1)
y = df['Appliances']

#Dividir os dados
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

#Treinar o modelo
rf.fit(x_train, y_train)

#Avalir o modelo
y_pred = rf.predict(x_test)

#R², RMSE e MAE
print('R²:', r2_score(y_test, y_pred))
print('RMSE:', np.sqrt(mean_squared_error(y_test, y_pred)))
print('MAE:', mean_absolute_error(y_test, y_pred))

R²: 0.5595510410669535
RMSE: 67.36303874044403
MAE: 32.30969430839385


## Utlizando Random Forest

In [20]:
# Random Forest
rf = RandomForestRegressor(n_estimators=100, random_state=42)

# Treinar o modelo
rf.fit(x_train, y_train)

# Avaliar o modelo
y_pred = rf.predict(x_test)

# R², RMSE e MAE
print('R²:', r2_score(y_test, y_pred))
print('RMSE:', np.sqrt(mean_squared_error(y_test, y_pred)))
print('MAE:', mean_absolute_error(y_test, y_pred))

R²: 0.563526818977582
RMSE: 67.05831829604588
MAE: 32.16286100320892


# **Comparação de Modelos e Conclusão**

Com base nos resultados que obtivemos para a Regressão Linear e o Random Forest, podemos comparar o desempenho dos modelos:

**Regressão Linear:**
R²: 0.1489
RMSE: 93.64
MAE: 53.64

**Random Forest Regressor:**
R²: 0.5635
RMSE: 67.06
MAE: 32.16

**Discussão:**

O coeficiente de determinação (R²) indica a proporção da variância na variável dependente que é previsível a partir das variáveis independentes. Um R² mais alto significa que o modelo explica uma maior parte da variação. Neste caso, o Random Forest Regressor tem um R² significativamente maior (0.5635) em comparação com a Regressão Linear (0.1489). Isso sugere que o Random Forest explica uma proporção muito maior da variação no consumo de energia ('Appliances').

O Erro Quadrático Médio da Raiz (RMSE) e o Erro Médio Absoluto (MAE) são métricas que medem a magnitude dos erros do modelo. Valores mais baixos de RMSE e MAE indicam um melhor desempenho. O Random Forest Regressor apresenta valores de RMSE (67.06) e MAE (32.16) consideravelmente menores do que a Regressão Linear (RMSE: 93.64, MAE: 53.64). Isso significa que, em média, as previsões do Random Forest estão mais próximas dos valores reais de consumo de energia.

**Conclusão:**

Com base nas métricas de avaliação (R², RMSE e MAE), o modelo **Random Forest Regressor** explica melhor o consumo de energia ('Appliances') do que o modelo de Regressão Linear para este conjunto de dados. Ele captura uma maior proporção da variância e tem erros de previsão menores.